### PR TITLE
discovery::server: fix activeUser field of getInfo

### PR DESCRIPTION
**Pull request by dsheets**
The current active user was almost persisted and reported but wasn't. The handler for the discovery server is under an Arc so I introduced a Mutex to protect the username field and made setting the username and sending the credentials atomic so that every valid credential send corresponds to an active user update.

There doesn't appear currently to be any way for this `activeUser` field to be reset back to the empty string when a user disconnects so one could argue that it's actually more broken now than before as it will keep reporting the last user to connect indefinitely instead of the user with a currently active session.

If you're interested in this fix, I may take a look at clearing the value in the future.

This issue was reported by @TimotheeGerber in <https://github.com/TimotheeGerber/spotify-connect/pull/4#discussion_r1428834453>.

### CODE REVIEW

**Comment by roderickvd** [comment]
Thanks! Yes this is a known shortcoming, nice to see something being done about it.

> If you're interested in this fix, I may take a look at clearing the value in the future.

That *would* be nice. Do you want to handle that under this PR or a separate one?

**Comment by roderickvd** [comment]
*bump* you think you can take a shot at updating the user on disconnect / reconnect?

**Comment by dsheets** [comment]
Hi, sorry for the delay. I might have some time to take a look at this next week.

**Comment by Finomnis** [comment]
Don't use async locks for short-lived locks like this. Use a normal `std` mutex instead. See [this section of the tokio book](https://tokio.rs/tokio/tutorial/shared-state#on-using-stdsyncmutex) for more information.

**Comment by roderickvd** [comment]
*bump* @Finomnis is right. Could you change it accordingly?

File: discovery/src/server.rs [hunk 1]
```diff
@@ -12,7 +12,7 @@ use aes::cipher::{KeyIvInit, StreamCipher};
 use base64::engine::general_purpose::STANDARD as BASE64;
 use base64::engine::Engine as _;
 use futures_core::Stream;
-use futures_util::{FutureExt, TryFutureExt};
+use futures_util::{lock::Mutex, FutureExt, TryFutureExt};
```
Comment (roderickvd) [inline discovery/src/server.rs:15]: Please use `Mutex` from `std` instead of `async` muteness.

File: discovery/src/server.rs [hunk 2]
```diff
@@ -55,21 +55,19 @@ impl RequestHandler {
...
-    fn handle_get_info(&self) -> Response<hyper::Body> {
+    async fn handle_get_info(&self) -> Response<hyper::Body> {
         let public_key = BASE64.encode(self.keys.public_key());
         let device_type: &str = self.config.device_type.into();
-        let mut active_user = String::new();
-        if let Some(username) = &self.username {
-            active_user = username.to_string();
-        }
+        let maybe_username = self.username.lock().await;
+        let active_user = maybe_username.as_deref().unwrap_or("");
```
Comment (roderickvd) [inline discovery/src/server.rs:70]: Can you comment why the `unwrap_or` is "safe" here? Should it always succeed?

**Review by roderickvd — CHANGES_REQUESTED** [review]

**Comment by roderickvd** [comment]
Nudging [dsheets](https://github.com/dsheets) - this would be a very worthwhile addition and only requires a bit of work... could you muster it and add to the changelog? Then we can merge!

**Comment by roderickvd** [comment]
We're aiming for a 0.5 release next week. Would be great to get this in, if you could do the last touch-up. Thanks.

File: discovery/src/server.rs (hunk 1 above)
Comment (dsheets) [reply discovery/src/server.rs]: Done.

File: discovery/src/server.rs (hunk 2 above)
Comment (dsheets) [reply discovery/src/server.rs]: The behavior is the same as previously, though the fixup modifies this hunk. There is no panic risk.

**Comment by dsheets** [comment]
Sorry for the absurd delay. I've changed the lock to `std` and rebased on `dev`.

**Comment by roderickvd** [comment]
Thanks!

**Pull request merged**